### PR TITLE
[feature](multi-catalog) support Jdbc catalog

### PR DESCRIPTION
**Pull request by BePPPower**
# Proposed changes

Issue Number: close #xxx

I add `jdbc catalog` for doris multi-catalog feature.
Currently, the `jdbc catalog` only supports `MYSQL` DBMS.

TODO:
1. support for `postgre` DB
2. Support for other databases.

## Problem summary

For `jdbc catalog`, we can create catalog like:
```sql
CREATE CATALOG jdbc4 PROPERTIES (
    "type"="jdbc",
    "jdbc.user"="root",
    "jdbc.password"="123456",
    "jdbc.jdbc_url" = "jdbc:mysql://127.0.0.1:13396/demo?yearIsDateType=false",
    "jdbc.driver_url" = "file:/mnt/disk2/ftw/tools/jar/mysql-connector-java-5.1.47/mysql-connector-java-5.1.47.jar",
    "jdbc.driver_class" = "com.mysql.jdbc.Driver"
);
```

Note:
`yearIsDateType` is a param of jdbc:
If yearIsDateType configuration property is set to false, then the returned object type is java.sql.Short. If set to true (the default), then the returned object is of type java.sql.Date with the date set to January 1st, at midnight.
To compat with mysql, we force the use of `yearIsDateType=false` in `FE`. if user sets `yearIsDateType=true`, doris FE will force to change `yearIsDateType=false`.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
3. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
5. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
6. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.06 seconds
 load time: 440 seconds
 storage size: 17170616360 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221123130823_clickbench_pr_51640.html

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+        this.driverClass = driverClass;
+        this.dbType = parseDbType(jdbcUrl);
+        this.checkSum = computeObjectChecksum();
+
+        try {
+            URL[] urls = {new URL(driverUrl)};
+            // set parent ClassLoader to null, we can achieve class loading isolation.
+            classLoader = URLClassLoader.newInstance(urls, null);
+        } catch (Exception e) {
+            throw new JdbcClientException("failed to load jar");
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java:87]: add detail error msg

File: fe/fe-core/src/main/java/org/apache/doris/datasource/JdbcExternalCatalog.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        return jdbcClient.isTableExist(dbName, tblName);
+    }
+
+    @Override
+    public void gsonPostProcess() throws IOException {
+        super.gsonPostProcess();
+        setProperties(this.catalogProperty.getProperties());
+    }
+
+    // TODO(ftw): 每次调用getSchema都会建立jdbc conn，会不会性能差？
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/JdbcExternalCatalog.java:140]: This is only called when schema cache is missing, so I think it is ok

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+            case "DATE":
+                return ScalarType.getDefaultDateType(Type.DATE);
+            case "TIMESTAMP":
+                return ScalarType.getDefaultDateType(Type.DATETIME);
+            case "DATETIME":
+                return ScalarType.createCharType(8);
+            case "TIME":
+                return ScalarType.createCharType(3);
+            case "YEAR":
+                return ScalarType.createCharType(1);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java:342]: For YEAR/TIME/TIMESTAMP, maybe  we should use integer. Better to test them.

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+            case "FLOAT":
+                return Type.FLOAT;
+            case "DOUBLE":
+                return Type.DOUBLE;
+            case "DECIMAL":
+                int precision = fieldSchema.getColumnSize();
+                // TODO(ftw): check
+                int scale = fieldSchema.getDecimalDigits();
+                return ScalarType.createDecimalType(precision, scale);
+            case "CHAR":
```
[the remark this answers is not in the document]
Comment (BePPPower) [reply fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java:352]: Why do we use doris String for mysql char type?  Isn't that a waste of space?

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        switch (dbType) {
+            case MYSQL:
+                return mysqlTypeToDoris(fieldSchema);
+            default:
+                throw new JdbcClientException("Unknown database type");
+        }
+    }
+
+    public Type mysqlTypeToDoris(JdbcFieldSchema fieldSchema) {
+        // For mysql type: "INT UNSIGNED", we use split(" ")[0] to get "INT"
```
[the remark this answers is not in the document]
Comment (BePPPower) [reply fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java (hunk 3 above)
Comment (BePPPower) [reply fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java]: After test:
for YEAR, we can use doris MEDIUMINT data type.
for TIME, we can use doris STRING data type.
for TIMESTAMP, we can use doris DATETIME data type.

File: fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+            URL[] urls = {new URL(driverUrl)};
+            // set parent ClassLoader to null, we can achieve class loading isolation.
+            classLoader = URLClassLoader.newInstance(urls, null);
+        } catch (MalformedURLException e) {
+            throw new JdbcClientException("MalformedURLException to load class about " + driverUrl, e);
+        }
+    }
+
+    public String parseDbType(String url) {
+        if (url.startsWith("jdbc:mysql")) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java:91]: jdbc:mariadb://[127.0.0.1](http://127.0.0.1/):%d/%s

Comment (BePPPower) [reply fe/fe-core/src/main/java/org/apache/doris/external/jdbc/JdbcClient.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/datasource/JdbcExternalCatalog.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        this.name = name;
+        this.type = "jdbc";
+        setProperties(props);
+        this.catalogProperty = new CatalogProperty();
+        this.catalogProperty.setProperties(props);
+    }
+
+    @Override
+    public void onClose() {
+        jdbcClient.closeClient();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/JdbcExternalCatalog.java:66]: jdbcClient may not be set, should check `if (jdbcClient != null)`

Comment (BePPPower) [reply fe/fe-core/src/main/java/org/apache/doris/datasource/JdbcExternalCatalog.java]: done

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**